### PR TITLE
avoid unneeded recomputing for 'on canvas' only changes

### LINKED ISSUE #11367: Crop recompute is triggered too often which makes it unresponsive
**Describe the bug/issue**
When I activate the 'crop' module, Darktable displays 'working'.
Whenever I resize, it displays 'working'. 
Every action triggers a few seconds delay.

It's probably my lack of understanding but as I see it, it knows all the pixels already. They are not going to be different.
This 'working' seems to be the entire pipeline which recomputes the entire image.

The problem is that the crop module becomes very unresponsive until it's done 'working'.  One issue I face very often is that it fails to switch from 'move' to 'resize' mode because it was 'working' when I moved the cursor. 

**To Reproduce**
Just crop a picture 

**Expected behavior**
I would like it to resize quickly and respond quickly to small adjustments and mouse movements.

### PR DESCRIPTION
**Pull request by AlicVB**
this fix #11367 

- First commit is a "massive" but quite simple change : it adds a "recompute" option to the function to add an history item. This allow to call this function without recomputing the pipe, which is required by next points
- adapt crop code in order to avoid pipe recompute when the module is focused, as in this case, the module is bypassed by the pipe anyway.
- adapt ashift code too, but this case is more complex, as some actions need pipe recompute (rotation for example) but some other don't (lines displaying, crop change, etc...)

@dterrahe : I'd like to have your review, especially on the ashift commit : I've removed the declaration of the crop combo with `dt_bauhaus_combobox_from_params` in favor of a manual one, otherwise pipe seems to be automatically recomputed... Thanks !

In my tests, I've not spotted any regression, but please test carefully !

Note that this changes may also be used for other modules (I've just tested crop and ashift for now)

### CODE REVIEW

File: src/iop/ashift.c [hunk 1]
```diff
@@ -5254,7 +5268,7 @@ static int _event_structure_auto_clicked(GtkWidget *widget, GdkEventButton *even
       g->jobparams = enhance;
     }
 
-    dt_dev_add_history_item(darktable.develop, self, TRUE, TRUE); // also calls dt_control_queue_redraw_center
+    dt_dev_add_history_item(darktable.develop, self, TRUE, FALSE); // also calls dt_control_queue_redraw_center
```
Comment (ralfbrown) [inline src/iop/ashift.c:5271]: Does the comment still apply when called with invalidate=FALSE?

Comment (AlicVB) [reply src/iop/ashift.c:5271]: yes, only the pipe recompute is "skipped". everything else remains the same

**Comment by dterrahe** [comment]
After having given this some thought, answering your question first:
if a bauhaus widget has a linked field (`_from_param` or `dt_bauhaus_widget_set_field`) any change will trigger `dt_iop_gui_changed` which calls `dt_dev_add_history_item` with `invalidate = TRUE`. The same is the case for the hidden "margins" sliders in the crop module, so those also still trigger recomputes in your PR, even when crop is focused. We could call `dt_bauhaus_widget_set_field` with `NULL` to break the link (this would need a small fix to avoid a warning), but...

I would have implemented this simpler; setting a new field in `dev` to indicate if a history change should trigger a recompute immediately and setting this to FALSE in `gui_focus` (`in` == TRUE) for just crop and ashift. Then, in those modules, for the few cases where you _do_ need to recompute, call `dt_dev_invalidate_all` explicitly.

Looking at crop and shift, I wonder if the whole shadow crop box concept (with calls to `_commit(_crop)_box`) is still needed or if this only served to avoid this same issue. If not needed, they should be cleaned up _now_ or a future observer will be lost and we end up exactly with the same unmaintainable module that we had in clipping.c

Other thought, probably more for @TurboGit who has been working on this so much recently; currently the second preview window also "zooms out" when the crop module gets the focus. Since it is not being used to _adjust_ the crop, would it not make more sense to show the actual result always? Or would having a different zoom there add too much complexity to the pipes? If that could work, then obviously preview2 would _always_ need to be invalidated by a history change.

File: src/iop/ashift.c [hunk 2]
```diff
@@ -5046,6 +5046,12 @@ void gui_reset(struct dt_iop_module_t *self)
   dt_dev_pixelpipe_flush_caches(self->dev->preview_pipe);
 }
 
+static int _gui_has_focus(struct dt_iop_module_t *self)
```
Comment (dterrahe) [inline src/iop/ashift.c:5049]: There's already a `gui_has_focus` that seems to be identical.

**Comment by AlicVB** [comment]
closing in favor of #13274

**Pull request closed without merging**